packages/profile-sync-controller/src/AuthenticationController.ts

```diff
@@ -0,0 +1,338 @@
+import type { UserStorageControllerDisableProfileSyncing } from './UserStorageController';
+
+const THIRTY_MIN_MS = 1000 * 60 * 30;
+
+const controllerName = 'AuthenticationController';
+
+type SessionProfile = {
+  identifierId: string;
+  profileId: string;
+};
+
+type SessionData = {
+  /** profile - anonymous profile data for the given logged in user */
+  profile: SessionProfile;
+  /** accessToken - used to make requests authorized endpoints */
+  accessToken: string;
+  /** expiresIn - string date to determine if new access token is required  */
+  expiresIn: string;
+};
+
+type MetaMetricsAuth = {
+  getMetaMetricsId: () => string;
+};
+
+export type AuthenticationControllerState = {
+  /**
+   * Global isSignedIn state.
+   * Can be used to determine if "Profile Syncing" is enabled.
+   */
+  isSignedIn: boolean;
+  sessionData?: SessionData;
+  internalAccounts?: {
+    accounts: Record<string, InternalAccount>;
+    selectedAccount: string;
+  };
+};
+const defaultState: AuthenticationControllerState = { isSignedIn: false };
+const metadata: StateMetadata<AuthenticationControllerState> = {
+  isSignedIn: {
+    persist: true,
+    anonymous: true,
+  },
+  sessionData: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+// Messenger Actions
+type CreateActionsObj<T extends keyof AuthenticationController> = {
+  [K in T]: {
+    type: `${typeof controllerName}:${K}`;
+    handler: AuthenticationController[K];
+  };
+};
+type ActionsObj = CreateActionsObj<
+  | 'performSignIn'
+  | 'performSignOut'
+  | 'getBearerToken'
+  | 'getSessionProfile'
+  | 'isSignedIn'
+>;
+export type Actions = ActionsObj[keyof ActionsObj];
+export type AuthenticationControllerPerformSignIn = ActionsObj['performSignIn'];
+export type AuthenticationControllerPerformSignOut =
+  ActionsObj['performSignOut'];
+export type AuthenticationControllerGetBearerToken =
+  ActionsObj['getBearerToken'];
+export type AuthenticationControllerGetSessionProfile =
+  ActionsObj['getSessionProfile'];
+export type AuthenticationControllerIsSignedIn = ActionsObj['isSignedIn'];
+
+// Allowed Actions
+export type AllowedActions =
+  | HandleSnapRequest
+  | UserStorageControllerDisableProfileSyncing;
+
+// Messenger
+export type AuthenticationControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  Actions | AllowedActions,
+  never,
+  AllowedActions['type'],
+  never
+>;
+
+/**
+ * Controller that enables authentication for restricted endpoints.
+ * Used for Global Profile Syncing and Notifications
+ */
+export class AuthenticationController extends BaseController<
+  typeof controllerName,
+  AuthenticationControllerState,
+  AuthenticationControllerMessenger
+> {
+  #metametrics: MetaMetricsAuth;
+
+  constructor({
+    messenger,
+    state,
+    metametrics,
+  }: {
+    messenger: AuthenticationControllerMessenger;
+    state?: AuthenticationControllerState;
+    /**
+     * Not using the Messaging System as we
```

Comment:
I don't understand this comment. The messaging system is used by mobile as well. Can you elaborate?

Reply:
I can answer this. At the time we split this up since we didn't see if there was a shared metametrics controller on both extension and mobile.

Extension uses this [`metametrics.js`](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/metametrics.js) controller.

Mobile I believe uses [this](https://github.com/MetaMask/metamask-mobile/blob/main/app/core/Analytics/MetaMetrics.ts).

If/when both use the same metametrics and use the same messaging system actions, then we can utilise the messaging system. For now we'll keep it an interface that you can pass in the implementation via the constructor.